File: packages/react-native-messaging-core/src/types/Conversation.ts
import { ProviderInConversation } from './ProviderInConversation';
import { ConversationId } from './ConversationId';
import { ConversationMessage } from "./ConversationItem";

export class Conversation {
  id: ConversationId;
  title?: string;
  inboxPreviewTitle: string;
  lastMessagePreview?: string;
  lastModified: Date;
  patientUnreadMessageCount: number;
  pictureURL?: string;
  providers: ProviderInConversation[];
  isLocked: boolean;
  lastMessage?: ConversationMessage;

  constructor(
    id: ConversationId,
    inboxPreviewTitle: string,
    lastModified: Date,
    patientUnreadMessageCount: number,
    providers: ProviderInConversation[],
    isLocked: boolean,
    title?: string,
    lastMessagePreview?: string,
    pictureURL?: string,
    lastMessage?: ConversationMessage,
  ) {
    this.id = id;
    this.title = title;
    this.inboxPreviewTitle = inboxPreviewTitle;
    this.lastMessagePreview = lastMessagePreview;
    this.lastModified = lastModified;
    this.patientUnreadMessageCount = patientUnreadMessageCount;
    this.pictureURL = pictureURL;
    this.providers = providers;
    this.isLocked = isLocked;
    this.lastMessage = lastMessage;
  }
}

File: packages/react-native-messaging-core/android/src/main/java/com/nabla/sdk/reactnative/messaging/core/models/Conversation.kt
package com.nabla.sdk.reactnative.messaging.core.models

import android.content.Context
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.nabla.sdk.core.annotation.NablaInternal
import com.nabla.sdk.core.domain.entity.evaluate
import com.nabla.sdk.messaging.core.domain.entity.Conversation

@JvmName("toConversationMapArray")
internal fun List<Conversation>.toMapArray(context: Context): ReadableArray = Arguments.createArray().apply {
    forEach { conversation ->
        pushMap(conversation.toMap(context))
    }
}

@OptIn(NablaInternal::class)
internal fun Conversation.toMap(context: Context): ReadableMap {
    return Arguments.createMap().apply {
        putMap("id", id.toMap())
        putString("title", title)
        putString("inboxPreviewTitle", inboxPreviewTitle.evaluate(context))
        putString("lastMessagePreview", lastMessagePreview)
        putString("lastModified", lastModified.toString())
        putInt("patientUnreadMessageCount", patientUnreadMessageCount)
        pictureUrl?.url?.toString()?.let { putString("pictureURL", it) }
        putArray("providers", providersInConversation.toMapArray())
        putBoolean("isLocked", isLocked)
    }
}

